static/test/javascripts/spec/common/commercial/dfp/messenger.spec.js

```diff
@@ -0,0 +1,134 @@
+            onMessage({ origin: 'http://google.com', source: mockFrame });
+            expect(mockFrame.postMessage).toHaveBeenCalled();
+            expect(response.error.code).toBe(400);
+        });
+
+        it('should respond with a 418 code when sending malformed JSON', function () {
+            onMessage({ origin: 'http://localhost:9876', data: '{', source: mockFrame });
+            expect(mockFrame.postMessage).toHaveBeenCalled();
+            expect(response.error.code).toBe(418);
+        });
+
+        it('should respond with a 400 code when sending incomplete payload', function () {
+            var payloads = [
+                { type: 'missing data' },
+                { value: 'missing type' },
+                { type: 'unregistered', value: 'type' }
+            ];
+            onMessage({ origin: 'http://localhost:9876', data: JSON.stringify(payloads[0]), source: mockFrame });
+            expect(mockFrame.postMessage).toHaveBeenCalled();
+            expect(response.error.code).toBe(400);
+            onMessage({ origin: 'http://localhost:9876', data: JSON.stringify(payloads[1]), source: mockFrame });
+            expect(mockFrame.postMessage).toHaveBeenCalled();
+            expect(response.error.code).toBe(400);
+            onMessage({ origin: 'http://localhost:9876', data: JSON.stringify(payloads[2]), source: mockFrame });
+            expect(mockFrame.postMessage).toHaveBeenCalled();
+            expect(response.error.code).toBe(400);
+        });
+
+        it('should respond with a 405 code when no listener is attached to a message type', function () {
+            var payload = { type: 'that', value: 'hello' };
+            messenger.register('this', routines.noop, mockWindow);
+            messenger.register('that', routines.noop, mockWindow);
+            messenger.unregister('that', routines.noop, mockWindow);
+            onMessage({ origin: 'http://localhost:9876', data: JSON.stringify(payload), source: mockFrame });
+            expect(mockFrame.postMessage).toHaveBeenCalled();
+            expect(response.error.code).toBe(405);
+            expect(response.error.message).toBe('Service that not implemented');
+            messenger.unregister('this', routines.noop, mockWindow);
+        });
+
+        it('should throw when the listener fails', function (done) {
+            var payload = { type: 'this', value: 'hello' };
+            messenger.register('this', routines.thrower, mockWindow);
+            onMessage({ origin: 'http://localhost:9876', data: JSON.stringify(payload), source: mockFrame })
+            .then(function () {
+                expect(mockFrame.postMessage).toHaveBeenCalled();
+                expect(response.error.code).toBe(500);
+                expect(response.error.message).toBe('Internal server error\n\ncatch this if you can!');
+                messenger.unregister('this', routines.thrower, mockWindow);
+            })
+            .then(done);
```

Comment:
you can add a `.catch(done.fail)` so if your expectation fails your test end immediately instead of timing out